ksql-engine/src/main/java/io/confluent/ksql/util/AvroUtil.java

```diff
@@ -0,0 +1,174 @@
+        int schemaId = -1;
+        try {
+          schemaId = Integer.parseInt(StringUtil.cleanQuotes(abstractStreamCreateStatement.getProperties().get(KsqlConstants.AVRO_SCHEMA_ID).toString()));
+        } catch (NumberFormatException e) {
+          throw new KsqlException(String.format("Invalid schema id property: %s.",
+                                                abstractStreamCreateStatement.getProperties().get(KsqlConstants.AVRO_SCHEMA_ID).toString()));
+        }
+        schemaMetadata = schemaRegistryClient.getSchemaMetadata(
+            kafkaTopicName +KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX, schemaId);
+      } else {
+        schemaMetadata = schemaRegistryClient.getLatestSchemaMetadata(kafkaTopicName +
+                                                                      KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX);
+      }
+
+      String avroSchemaString = schemaMetadata.getSchema();
+      streamsProperties.put(DdlConfig.AVRO_SCHEMA, avroSchemaString);
+      Schema schema = SerDeUtil.getSchemaFromAvro(avroSchemaString);
+      if (abstractStreamCreateStatement.getElements().isEmpty()) {
+        abstractStreamCreateStatement = addAvroFields(abstractStreamCreateStatement, schema,
+                                                      schemaMetadata.getId());
+        return new Pair<>(abstractStreamCreateStatement, SqlFormatter.formatSql(abstractStreamCreateStatement));
+      } else {
+        return new Pair<>(abstractStreamCreateStatement, null);
+      }
+    } catch (Exception e) {
+      String errorMessage = String.format(" Could not fetch the AVRO schema from schema "
+                                          + "registry. %s ", e.getMessage());
+      throw new KsqlException(errorMessage);
+    }
+  }
+
+  private static AbstractStreamCreateStatement addAvroFields(AbstractStreamCreateStatement
+                                                                             abstractStreamCreateStatement, Schema
+                                                                             schema, int schemaId) {
+    List<TableElement> elements = new ArrayList<>();
+    for (Field field: schema.fields()) {
+      TableElement tableElement = new TableElement(field.name(), SchemaUtil
+          .getSQLTypeName(field.schema()));
+      elements.add(tableElement);
+    }
+    StringLiteral schemaIdLiteral = new StringLiteral(String.format("%d", schemaId));
+    Map<String, Expression> properties;
+    if (abstractStreamCreateStatement.getProperties().containsKey(KsqlConstants.AVRO_SCHEMA_ID)) {
+      properties = abstractStreamCreateStatement.getProperties();
+    } else {
+      properties = new HashMap<>();
+      properties.putAll(abstractStreamCreateStatement.getProperties());
+      properties.put(KsqlConstants.AVRO_SCHEMA_ID, schemaIdLiteral);
+    }
+
+
+    if (abstractStreamCreateStatement instanceof CreateStream) {
```

Comment:
can we not add an abstract method to `AbstracStreamCreateStatement`, i.e, `copyWith(List elements, Map properties)`
It is implemented by `CreateStream` and `CreateTable`. We can then remove this `if(..) else...`

Reply:
Good point. Made the changes as you suggested.